File: JTMaterialTransition/JTMaterialTransition.m
//
//  JTMaterialTransition.m
//  JTMaterialTransition
//
//  Created by Jonathan Tribouharet
//

#import "JTMaterialTransition.h"

@implementation JTMaterialTransition

- (instancetype)initWithAnimatedView:(UIView *)animatedView
{
    self = [super init];
    if(!self){
        return nil;
    }
    
    NSAssert(animatedView != nil, @"animatedView cannot be nil");
    _animatedView = animatedView;
    
    return self;
}

- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext
{
    return .5;
}

- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext
{
    if(_animatedView){
        NSAssert(_animatedView.superview != nil, @"animatedView must be attached to a superview");

        // Get the frame rect in the screen coordinates
        self.startFrame = [_animatedView.superview convertRect:_animatedView.frame toView:nil];
        self.startBackgroundColor = _animatedView.backgroundColor;
    }
    
    // Use if the transitionContext.containerView is not fullscreen
    CGRect startFrame = [transitionContext.containerView.superview convertRect:self.startFrame toView:transitionContext.containerView];
    
    UIView *animatedViewForTransition;
    {
        animatedViewForTransition = [[UIView alloc] initWithFrame:startFrame];
        [transitionContext.containerView addSubview:animatedViewForTransition];
        
        animatedViewForTransition.clipsToBounds = YES;
        animatedViewForTransition.layer.cornerRadius = CGRectGetHeight(animatedViewForTransition.frame) / 2.;
        animatedViewForTransition.backgroundColor = self.startBackgroundColor;
    }
    
    CGAffineTransform finalTransform;
    {
        CGFloat size = MAX(CGRectGetHeight(transitionContext.containerView.frame), CGRectGetWidth(transitionContext.containerView.frame)) * 1.2;
        CGFloat scaleFactor = size / CGRectGetWidth(animatedViewForTransition.frame);
        finalTransform = CGAffineTransformMakeScale(scaleFactor, scaleFactor);
    }
    
    UIViewController *presentedController;
    if(!self.isReverse){
        presentedController = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
        
        presentedController.view.frame = transitionContext.containerView.bounds;
        presentedController.view.layer.opacity = 0;
        [transitionContext.containerView addSubview:presentedController.view];
    }
    else{
        presentedController = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
        
        presentedController.view.frame = transitionContext.containerView.bounds;
        [transitionContext.containerView addSubview:presentedController.view];
    }
    
    if(!self.isReverse){
        [UIView transitionWithView:animatedViewForTransition
                          duration:[self transitionDuration:transitionContext] * .7
                           options:0
                        animations:^{
                            animatedViewForTransition.transform = finalTransform;
                            animatedViewForTransition.center = transitionContext.containerView.center;
                            animatedViewForTransition.backgroundColor = presentedController.view.backgroundColor;
                        } completion:nil];
        
        [UIView animateWithDuration:[self transitionDuration:transitionContext] * .42
                              delay:[self transitionDuration:transitionContext] * .58
                            options:0
                         animations:^{
                             presentedController.view.layer.opacity = 1;
                         } completion:^(BOOL finished) {
                             [animatedViewForTransition removeFromSuperview];
                             [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
                         }];
    }
    else{
        animatedViewForTransition.transform = finalTransform;
        animatedViewForTransition.center = transitionContext.containerView.center;
        animatedViewForTransition.backgroundColor = presentedController.view.backgroundColor;
        
        [UIView animateWithDuration:[self transitionDuration:transitionContext] * .7
                         animations:^{
                             presentedController.view.layer.opacity = 0;
                         } completion:nil];
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * [self transitionDuration:transitionContext] * .32), dispatch_get_main_queue(), ^{
            [UIView transitionWithView:animatedViewForTransition
                              duration:[self transitionDuration:transitionContext] * .58
                               options:0
                            animations:^{
                                animatedViewForTransition.transform = CGAffineTransformIdentity;
                                animatedViewForTransition.backgroundColor = self.startBackgroundColor;
                                animatedViewForTransition.frame = startFrame;
                            } completion:^(BOOL finished) {
                                [animatedViewForTransition removeFromSuperview];
                                [transitionContext completeTransition:![transitionContext transitionWasCancelled]];
                            }];
        });
    }
}

@end

File: JTMaterialTransition/JTMaterialTransition.swift
//
//  JTMaterialTransition.swift
//  JTMaterialTransition
//
//  Created by Jonathan Tribouharet
//

import UIKit

open class JTMaterialTransition: NSObject, UIViewControllerAnimatedTransitioning, UIViewControllerTransitioningDelegate {
    
    open weak var animatedView: UIView?
    open var startFrame = CGRect()
    open var startBackgroundColor: UIColor?
    open var isReverse = false
    open var transitionDuration : TimeInterval = 0.5

    convenience public init(animatedView: UIView) {
        self.init()
        self.animatedView = animatedView
    }
    
    //MARK - UIViewControllerAnimatedTransitioning
    
    open func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return self.transitionDuration
    }
    
    open func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        if let animatedView = animatedView, let superview = animatedView.superview {
            
            // Get the frame rect in the screen coordinates
            self.startFrame = superview.convert(animatedView.frame, to: nil)
            self.startBackgroundColor = animatedView.backgroundColor
        }
        
        // Use if the transitionContext.containerView is not fullscreen
        let startFrame = transitionContext.containerView.superview?.convert(self.startFrame, to: transitionContext.containerView) ?? self.startFrame
        
        let animatedViewForTransition = UIView(frame: startFrame)
        transitionContext.containerView.addSubview(animatedViewForTransition)
        
        animatedViewForTransition.clipsToBounds = true
        animatedViewForTransition.layer.cornerRadius = animatedViewForTransition.frame.height / 2.0
        animatedViewForTransition.backgroundColor = self.startBackgroundColor
        

        let presentedController: UIViewController
        
        if !self.isReverse {
            presentedController = transitionContext.viewController(forKey: .to)!
            presentedController.view.layer.opacity = 0
        }
        else {
            presentedController = transitionContext.viewController(forKey: .from)!
        }
        
        presentedController.view.frame = transitionContext.containerView.bounds
        transitionContext.containerView.addSubview(presentedController.view)
        
        
        let size = max(transitionContext.containerView.frame.height, transitionContext.containerView.frame.width) * 1.2
        let scaleFactor = size / animatedViewForTransition.frame.width
        let finalTransform = CGAffineTransform(scaleX: scaleFactor, y: scaleFactor)
        
        
        if !self.isReverse {
            UIView.transition(with: animatedViewForTransition,
                              duration: self.transitionDuration(using: transitionContext) * 0.7,
                              options: [],
                              animations: {
                                animatedViewForTransition.transform = finalTransform
                                animatedViewForTransition.center = transitionContext.containerView.center
                                animatedViewForTransition.backgroundColor = presentedController.view.backgroundColor
            },
                              completion: { (_) in
            })
            
            UIView.animate(withDuration: self.transitionDuration(using: transitionContext) * 0.42,
                           delay: self.transitionDuration(using: transitionContext) * 0.58,
                           animations: {
                            presentedController.view.layer.opacity = 1
            },
                           completion: { (_) in
                            animatedViewForTransition.removeFromSuperview()
                            transitionContext.completeTransition(!transitionContext.transitionWasCancelled)
            })
        }
        else {
            animatedViewForTransition.transform = finalTransform
            animatedViewForTransition.center = transitionContext.containerView.center
            animatedViewForTransition.backgroundColor = presentedController.view.backgroundColor
            
            UIView.animate(withDuration: self.transitionDuration(using: transitionContext) * 0.7, animations: {
                presentedController.view.layer.opacity = 0
            })
            
            DispatchQueue.main.asyncAfter(deadline: .now() + self.transitionDuration(using: transitionContext) * 0.32) {
                UIView.transition(with: animatedViewForTransition,
                                  duration: self.transitionDuration(using: transitionContext) * 0.58,
                                  options: [],
                                  animations: {
                                    animatedViewForTransition.transform = CGAffineTransform.identity
                                    animatedViewForTransition.backgroundColor = self.startBackgroundColor
                                    animatedViewForTransition.frame = startFrame
                },
                                  completion: { (_) in
                                    animatedViewForTransition.removeFromSuperview()
                                    transitionContext.completeTransition(!transitionContext.transitionWasCancelled)
                })
            }
        }
    }
    
    // MARK - UIViewControllerTransitioningDelegate
    
    open func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isReverse = false
        return self
    }
    
    open func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isReverse = true
        return self
    }
    
}
